// generator-angular-fullstack/generators/heroku/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Generator = undefined;

var _util = require('util');

var _util2 = _interopRequireDefault(_util);

var _yeomanGenerator = require('yeoman-generator');

var _yeomanGenerator2 = _interopRequireDefault(_yeomanGenerator);

var _child_process = require('child_process');

var _chalk = require('chalk');

var _chalk2 = _interopRequireDefault(_chalk);

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _underscore = require('underscore.string');

var _underscore2 = _interopRequireDefault(_underscore);

var _generatorBase = require('../generator-base');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

class Generator extends _yeomanGenerator.Base {
  constructor(...args) {
    super(...args);

    this.sourceRoot(_path2.default.join(__dirname, '../../templates/heroku'));
  }

  initializing() {
    return (0, _generatorBase.genBase)(this);
  }

  askForName() {
    var done = this.async();

    var prompts = [{
      name: 'deployedName',
      message: 'Name to deploy as (Leave blank for a random name):'
    }];

    return this.prompt(prompts).then(props => {
      this.deployedName = _underscore2.default.slugify(props.deployedName);
    });
  }

  askForRegion() {
    var done = this.async();

    var prompts = [{
      type: 'list',
      name: 'region',
      message: 'On which region do you want to deploy ?',
      choices: ['US', 'EU'],
      default: 0
    }];

    return this.prompt(prompts).then(props => {
      this.region = props.region.toLowerCase();
    });
  }

  checkInstallation() {
    if (this.abort) return;
    var done = this.async();

    (0, _child_process.exec)('heroku --version', err => {
      if (err) {
        this.log.error('You don\'t have the Heroku Toolbelt installed. Grab it from https://toolbelt.heroku.com/');
        this.abort = true;
      }
      done();
    });
  }

  gitInit() {
    if (this.abort) return;
    var done = this.async();

    this.log(_chalk2.default.bold('\nInitializing deployment repo'));
    this.mkdir('dist');
    var child = (0, _child_process.exec)('git init', { cwd: 'dist' }, (err, stdout, stderr) => {
      done();
    });

    child.stdout.on('data', data => {
      this.log(data.toString());
    });
  }

  herokuCreate() {
    if (this.abort) return;
    var done = this.async();
    var regionParams = this.region !== 'us' ? ' --region ' + this.region : '';

    this.log(_chalk2.default.bold('Creating heroku app and setting node environment'));
    var child = (0, _child_process.exec)(`heroku apps:create ${this.deployedName + regionParams} && heroku config:set NODE_ENV=production`, { cwd: 'dist' }, (err, stdout, stderr) => {
      if (err) {
        this.abort = true;
        this.log.error(err);
      } else {
        this.log('stdout: ' + stdout);
      }
      done();
    });

    child.stdout.on('data', data => {
      this.log(data.toString());
    });
  }

  copyProcfile() {
    if (this.abort) return;
    var done = this.async();
    this.log(_chalk2.default.bold('Creating Procfile'));
    this.copy('Procfile', 'dist/Procfile');
    this.conflicter.resolve(err => {
      done();
    });
  }

  build() {
    if (this.abort) return;
    var done = this.async();

    this.log(_chalk2.default.bold('\nBuilding dist folder, please wait...'));

    var child = (0, _child_process.exec)('gulp build', (err, stdout) => {
      done();
    });

    child.stdout.on('data', data => {
      this.log(data.toString());
    });
  }

  gitInit() {
    if (this.abort) return;
    var done = this.async();

    this.log(_chalk2.default.bold('Adding files for initial commit'));
    var child = (0, _child_process.exec)('git add -A && git commit -m "Initial commit"', { cwd: 'dist' }, function (err, stdout, stderr) {
      if (stdout.search('nothing to commit') >= 0) {
        this.log('Re-pushing the existing "dist" build...');
      } else if (err) {
        this.log.error(err);
      } else {
        this.log(_chalk2.default.green('Done, without errors.'));
      }
      done();
    }.bind(this));

    child.stdout.on('data', data => {
      this.log(data.toString());
    });
  }

  gitForcePush() {
    if (this.abort) return;
    var done = this.async();

    this.log(_chalk2.default.bold("\nUploading your initial application code.\n This may take " + _chalk2.default.cyan('several minutes') + " depending on your connection speed..."));

    var child = (0, _child_process.exec)('git push -f heroku master', { cwd: 'dist' }, (err, stdout, stderr) => {
      if (err) {
        this.log.error(err);
      } else {
        var hasWarning = false;

        if (this.filters.mongoose) {
          this.log(_chalk2.default.yellow(`
Because you're using mongoose, you must add mongoDB to your heroku app.
\tfrom \`/dist\`: ${_chalk2.default.bold('heroku addons:create mongolab')}
`));
          hasWarning = true;
        }

        let oauthMessage = strategy => _chalk2.default.yellow(`
You will need to set environment variables for ${strategy} auth. From \`/dist\`:
\t${_chalk2.default.bold(`heroku config:set ${strategy.toUpperCase()}_ID=appId`)}
\t${_chalk2.default.bold(`heroku config:set ${strategy.toUpperCase()}_SECRET=secret`)}
`);
        if (this.filters.facebookAuth) {
          this.log(oauthMessage('facebook'));
          hasWarning = true;
        }
        if (this.filters.googleAuth) {
          this.log(oauthMessage('google'));
          hasWarning = true;
        }
        if (this.filters.twitterAuth) {
          this.log(oauthMessage('twitter'));
          hasWarning = true;
        }

        this.log(_chalk2.default.green(`
Your app should now be live. To view it run
\t${_chalk2.default.bold('cd dist && heroku open')}`));

        if (hasWarning) {
          this.log(_chalk2.default.green('\nYou may need to address the issues mentioned above and restart the server for the app to work correctly.'));
        }

        this.log(_chalk2.default.yellow(`
After app modification run
\t${_chalk2.default.bold('gulp build')}
Then deploy with
\t${_chalk2.default.bold('gulp buildcontrol:heroku')}`));
      }
      done();
    });

    child.stdout.on('data', data => {
      this.log(data.toString());
    });
  }
}

exports.Generator = Generator;
module.exports = Generator;
